### DISCUSSION TITLE
[city] param in RootLayout

### DISCUSSION THREAD

**Discussion by Dezmonter** [vercel/next.js#71646] [Help]
### Summary

Please help me! My application structure:
![image](https://github.com/user-attachments/assets/fc7bdd8f-fdc3-4809-b581-8665949ee97f)

this my RootLayout:
![image](https://github.com/user-attachments/assets/0ffe5d72-aa44-4e36-9241-9da0f8114f86)

I want initialCity to contain the value from cookies if there is no param [city] in the URL. And if the URL contains param [city], then the value was taken from it and written to cookies.
How can I do this? I don't want to make 2 layouts for the sake of this, I probably need to get [city] in the RootLayout.

I beg you to help me. I haven't been able to solve this problem for 2 weeks now (

**Comment by Mr-Meshky** [comment]
you can use cookies-next library:

```jsx filename="layout.tsx"
import { cookies } from 'next/headers';
import { setCookie } from 'cookies-next';

export default function Layout({ children, params }: { children: React.ReactNode, params: { city: string } }) {
  const cityParam = params?.city;
  
  const city = cityParam || 'london';

  if (cityParam) {
    setCookie('city', cityParam, { maxAge: 60 * 60 * 24 * 7 });
  }

  return (
    <html lang="en">
      <body>
        {children}
      </body>
    </html>
  );
}

```

i hope it help

**Reply by Dezmonter** [reply]
> вы можете использовать библиотеку cookies-next:
> 
> ```js
> import { cookies } from 'next/headers';
> import { setCookie } from 'cookies-next';
> 
> export default function Layout({ children, params }: { children: React.ReactNode, params: { city: string } }) {
>   const cityParam = params?.city;
>   
>   const city = cityParam || 'london';
> 
>   if (cityParam) {
>     setCookie('city', cityParam, { maxAge: 60 * 60 * 24 * 7 });
>   }
> 
>   return (
>     <html lang="en">
>       <body>
>         {children}
>       </body>
>     </html>
>   );
> }
> ```
> 
> надеюсь это поможет

The main problem is that I can't get the city parameter in RootLayout(
Since layout is not in the [city] folder

**Comment by saeid1994** [comment]
Hello,I think this is what you want , yes?

1. Check if the city parameter exists in the URL.
2. If it does, set it as the initialCity and save it in cookies.
3. If it doesn’t, get initialCity from cookies.

So here’s how you can set this up:

1. Use cookies from Next.js to access and set cookie values.
2. Use headers from next/headers to get the current URL parameters.

this is my code  for your problem please check it carfully and tell me if you did not understand it 


```tsx
import { cookies } from 'next/headers';
import { usePathname, useRouter } from 'next/navigation';
import { CityProvider } from './CityProvider';
import Header from './Header';
import Footer from './Footer';
import { getCity } from './api/getCity';

export default async function RootLayout({ children }: { children: React.ReactNode }) {
  const pathname = usePathname();
  const cityCookie = cookies().get('city');
  const cityParam = pathname.split('/').pop(); // Assuming [city] is at the end of the path

  // Determine initialCity based on city param or cookie
  let initialCity;
  if (cityParam && cityParam !== cityCookie?.value) {
    // Fetch city data based on URL param
    initialCity = await getCity(cityParam);
    // Set city in cookies for subsequent visits
    cookies().set('city', cityParam);
  } else {
    // Use the cookie if there’s no city param in URL
    initialCity = cityCookie?.value ? await getCity(cityCookie.value) : await getCity('london');
  }

  return (
    <CityProvider initialCity={initialCity}>
      <Header />
      {children}
      <Footer />
    </CityProvider>
  );
}
```

**Reply by Dezmonter** [reply]
> Здравствуйте, я думаю, это то, что вы хотите, да?
> 
> 1. Проверьте, присутствует ли параметр города в URL-адресе.
> 2. Если это так, установите его в качестве initialCity и сохраните в файлах cookie.
> 3. Если это не так, получите initialCity из файлов cookie.
> 
> Вот как это можно настроить:
> 
> 1. Используйте файлы cookie из Next.js для доступа к значениям файлов cookie и их установки.
> 2. Используйте заголовки из next/headers для получения текущих параметров URL.
> 
> Это мой код для решения вашей проблемы. Пожалуйста, внимательно проверьте его и сообщите мне, если вы его не поняли.
> 
> ```tsx
> import { cookies } from 'next/headers';
> import { usePathname, useRouter } from 'next/navigation';
> import { CityProvider } from './CityProvider';
> import Header from './Header';
> import Footer from './Footer';
> import { getCity } from './api/getCity';
> 
> export default async function RootLayout({ children }: { children: React.ReactNode }) {
>   const pathname = usePathname();
>   const cityCookie = cookies().get('city');
>   const cityParam = pathname.split('/').pop(); // Assuming [city] is at the end of the path
> 
>   // Determine initialCity based on city param or cookie
>   let initialCity;
>   if (cityParam && cityParam !== cityCookie?.value) {
>     // Fetch city data based on URL param
>     initialCity = await getCity(cityParam);
>     // Set city in cookies for subsequent visits
>     cookies().set('city', cityParam);
>   } else {
>     // Use the cookie if there’s no city param in URL
>     initialCity = cityCookie?.value ? await getCity(cityCookie.value) : await getCity('london');
>   }
> 
>   return (
>     <CityProvider initialCity={initialCity}>
>       <Header />
>       {children}
>       <Footer />
>     </CityProvider>
>   );
> }
> ```

Thank you very much, this is a great option, but how to catch the dynamic parameter of the city? He won't be at the end of the road...

I can't do it all rootLayout client component (((

for example:

site.com/london/doctors
site.com/london/clinics

or

site.com/about
site.com
site.com/contacts

**Reply by saeid1994** [reply]
To access the [city] parameter in the root layout when it's outside of the [city] folder, you can use next/navigation's useSearchParams to retrieve query parameters if you modify the URL to include the city as a query parameter (e.g., /somepage?city=london).

However, if you'd like to keep the city parameter as part of the path (e.g., /london), you can use middleware to rewrite the URL to pass city as a query parameter, making it accessible across all components in your layout.

Here's how to set it up:

1- Add a middleware.ts file at the root of your app directory to rewrite the city path segment as a query parameter : 

```tsx
// middleware.ts
import { NextResponse } from 'next/server';
import type { NextRequest } from 'next/server';

export function middleware(request: NextRequest) {
  const { pathname } = request.nextUrl;

  // Check if the path has a city-like segment
  const cityMatch = pathname.match(/^\/([^/]+)/);
  if (cityMatch) {
    const city = cityMatch[1];
    const url = request.nextUrl.clone();
    url.pathname = '/';
    url.searchParams.set('city', city);
    return NextResponse.rewrite(url);
  }

  return NextResponse.next();
}

export const config = {
  matcher: '/:path*',
};
```
2- Access the city Parameter in Root Layout
Now that the city is rewritten as a query parameter, you can access it in your root layout using useSearchParams:

```tsx
// app/layout.tsx
import { cookies } from 'next/headers';
import { useSearchParams } from 'next/navigation';
import { CityProvider } from './CityProvider';
import Header from './Header';
import Footer from './Footer';
import { getCity } from './api/getCity';

export default async function RootLayout({ children }: { children: React.ReactNode }) {
  const searchParams = useSearchParams();
  const cityParam = searchParams.get('city');
  const cityCookie = cookies().get('city');

  // Determine initialCity based on city param or cookie
  let initialCity;
  if (cityParam && cityParam !== cityCookie?.value) {
    initialCity = await getCity(cityParam);
    cookies().set('city', cityParam);
  } else {
    initialCity = cityCookie?.value ? await getCity(cityCookie.value) : await getCity('london');
  }

  return (
    <CityProvider initialCity={initialCity}>
      <Header />
      {children}
      <Footer />
    </CityProvider>
  );
}
```

**Reply by Dezmonter** [reply]
> ```tsx
>  const cityParam = searchParams.get('city');
> ```

Once again I want to thank you, but I can only use useSearchParams in client components... and the rootLayout should not be entirely client component(
